rfd/0003-workspace-ram-management.md

```diff
@@ -0,0 +1,70 @@
+- No support for discard, which results in the sparse disk files always growing in size.
+- Weird performance characteristics and longer boot time[^3].
+
+### THP (Transparent Huge Pages)
+
+Will always be disabled by default in Hocus workspaces to increase the effectiveness of free page reporting. THP might be enabled on the host, but we recommend disabling it for production usage.
+
+### KSM (Kernel Same Page Merging)
+
+Workspaces running on a single machine will usually run the same kernel, IDE version, and edit the same code. This setup is ideal for KSM. We recommend enabling KSM on the host machine, but it's not a requirement for Hocus.
+
+### DAMON (Data Access MONitor)
+
+For free page reporting to be effective, the workspace needs the ability to drop page caches it doesn't need. By default, Linux is designed to consume as much RAM as possible and never release it back until there is some memory pressure. Hocus workspaces use DAMON RECLAIM, available since the 5.16 kernel[^4], to proactively reclaim unused page caches. In our testing, this releases most of the caches the VM uses, but as DAMON is a very new subsystem in the kernel, there is still a lot of room for improvements[^5].
+
+### PSI-based autosizing
+
+DAMON is able to reclaim unused caches but won't prevent the page cache from growing uncontrollably as the VM is rarely under high memory pressure. To ensure RAM is only allocated to the VM when it's actually needed, Hocus Workspaces ship [Sempai](https://github.com/facebookincubator/senpai), which is responsible for autosizing the system cgroup to ensure the VM is always under some kind of memory pressure. Sempai uses the PSI kernel mechanism to determine whether the workspace is under memory pressure and needs to increase/decrease its memory limits.
+
+### Memlocking critical binaries
+
+Hocus tries to reclaim as much of the page cache as possible and as quickly as possible. Executable code is located in the page cache, and as a result, a developer might lose access to their workspace under high memory pressure. To ensure they do not lose access to the workspace, we employ [prelockd](https://github.com/hakavlad/prelockd) to ensure critical system binaries like `sshd` are never evicted from the page cache.
+
+### Using swap on the host
+
+Hocus is designed to run a ton of workspaces on a single machine as the usage pattern of dev machines is high usage spikes followed by periods of inactivity. In the unlikely event multiple developers run multiple heavy jobs at the same time, exceeding the available memory on the host machine, the OOM killer will kill some workspaces. If that's unwanted, then we recommend having swap enabled on the host machine to swap out some of the workspaces and give KSM time to deduplicate the pages of the new jobs.
+
+### Guest kernel version
+
+For now, Hocus Workspaces will ship with the 6.2 kernel. This might be extended in the future down to the 5.16 kernel.
```

Comment:
"This might be extended in the future down to the 5.16 kernel." - I don't understand this sentence. Please rewrite.